[libc++] Define deque::__block_size inline

ldionne opened the pull request:
We noticed that __block_size was sometimes generating weak-defs that
were exported from programs. This is clearly unintended and unnecessary.
Defining the constant inline should allow it to be inlined in the code
all the time, which should help with that. It also simplifies the code.

frederick-vs-ja commented on libcxx/include/deque:
```diff
@@ -547,7 +543,7 @@ private:
   };
 
-  static const difference_type __block_size;
+  static _LIBCPP_CONSTEXPR const difference_type __block_size = __deque_block_size<_ValueType, _DiffType>::value;
```
Copy-pasta?
```suggestion
  static _LIBCPP_CONSTEXPR const difference_type __block_size = __deque_block_size<_Tp, _DiffType>::value;
```

ldionne committed:
[libc++] Define deque::__block_size inline

We noticed that __block_size was sometimes generating weak-defs that
were exported from programs. This is clearly unintended and unnecessary.
Defining the constant inline should allow it to be inlined in the code
all the time, which should help with that. It also simplifies the code.

ldionne commented:
Reverting to the original approach.

frederick-vs-ja commented:
I think we may need to move `__block_size` into function bodies (and repeat it many times) to support incomplete element types and be optimization-friendly.
